apps/controller/main.cpp

```diff
@@ -0,0 +1,186 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+
+#include "mbed.h"
+#include "rover_config.h"
+
+#define DEBUG
+
+static constexpr uint32_t MAX_UINT32               = 0xFFFFFFFF;
+static constexpr uint16_t SEND_INTERVAL_MS         = 100;
+static constexpr uint8_t DEBOUNCE_THRES            = 3;
+static constexpr uint8_t ANALOG_NUM_SAMPLE_AVERAGE = 5;
+
+DigitalIn joy_btn(JOY_BTN);
+DigitalIn pb_1(PB1);
+DigitalIn pb_2(PB2);
+DigitalIn sw_1a(SW_1A);
+DigitalIn sw_1b(SW_1B);
+DigitalIn sw_2a(SW_2A);
+DigitalIn sw_2b(SW_2B);
+
+static DigitalIn digital_inputs[] = {
+    joy_btn, pb_1, pb_2, sw_1a, sw_1b, sw_2a, sw_2b,
+};
+
+static constexpr uint8_t NUM_DIGITAL_IN = sizeof(digital_inputs) / sizeof(PinName);
+// Although for the test board, there is only 7 digital inputs, in the real boards
+// there can be more than 8 digital inputs, so using uint16_t here
+uint16_t digital_inputs_values                 = 0;
+uint16_t digital_inputs_curr_values            = 0;
+uint16_t digital_inputs_counts[NUM_DIGITAL_IN] = {0};
+
+AnalogIn joy_sm_x(JOY_SM_X);
+AnalogIn joy_sm_y(JOY_SM_Y);
+AnalogIn joy_x(JOY_X);
+AnalogIn joy_y(JOY_Y);
+AnalogIn pot_al(POT_AL);
+AnalogIn slide_pot_al(SLIDE_POT_AL);
+
+static AnalogIn analog_inputs[] = {
+    joy_sm_x, joy_sm_y, joy_x, joy_y, pot_al, slide_pot_al,
+};
+
+static constexpr uint8_t NUM_ANALOG_IN                                  = sizeof(analog_inputs) / sizeof(PinName);
+uint16_t analog_inputs_values[NUM_ANALOG_IN][ANALOG_NUM_SAMPLE_AVERAGE] = {0};
+uint16_t analog_inputs_values_idx[NUM_ANALOG_IN]                        = {0};
+uint32_t analog_inputs_sums[NUM_ANALOG_IN]                              = {0};
+
+static constexpr uint8_t NUM_BYTES_DIGITAL_IN    = sizeof(uint16_t);
+static constexpr uint8_t NUM_BYTES_PER_ANALOG_IN = sizeof(uint16_t);
+static constexpr uint16_t NUM_BYTES_ANALOG_IN    = sizeof(uint16_t) * NUM_ANALOG_IN;
+
+// Timer for checking interval between sending data to PC
+Timer timer;
+
+// Serial object for sending data to PC
+Serial pc(USBTX, USBRX);
+
+// Integrating debouncing is applied, not sure if this is necessary, needs testing
+void read_but_switches(uint16_t& digital_inputs_values, uint16_t& digital_inputs_curr_values,
+                       uint16_t digital_inputs_counts[NUM_DIGITAL_IN]) {
+  for (int i = 0; i < NUM_DIGITAL_IN; i++) {
+    int val = digital_inputs[i].read();  // val = 0 or 1
+    // if we read the same value as previous read, increment count
+    if ((digital_inputs_curr_values & 1 << i) == val) digital_inputs_counts[i]++;
```

Comment:
I'm not sure, but I think it should be (val << i) instead of just val for this comparison, because it seems like you're comparing a byte (which could have a single 1 in any position) to an int that can only be 1 or 0

Reply:
oh yeah dumb me, gonna change that